### ISSUE TITLE
YOLOv8 bounding box dimensions are 0 when running detection on an ONNX model with OpenCV CUDA build

### ISSUE THREAD

**Issue by Krugpelke** [opencv/opencv#23977]
### System Information

**Training (python)**
python 3.8
ultralytics 8.0.128
opencv-python 4.8.0.74
torch 1.8.0+cu111
onnx 1.14.0
onnxsim 0.4.33

**Inference (c++)**
opencv 4.8.0
cuda 11.2.2
cudnn 8.1.1.33

### Detailed description

I'm loading a simple yolov8 model exported as onnx for object detection. I have tested and confirmed that both the model and code are working correctly when opencv is built without cuda enabled, however, when running inference with a cuda build, interestingly the resulting bounding box coordinates and size are always 0, yet the score is correct. Also, I'm not getting any errors here.
Just to note, I've used the same cuda build to successfully run many other models (yolov5, yolov7 included), perhaps my issue lies somewhere in yolov8 incompatibility with these cuda or cudnn versions, which I can't upgrade that simply due to many different reasons.

### Steps to reproduce

Here's how I've done it, the model I've tested was simply made as such:
```python
from ultralytics import YOLO
model = YOLO('yolov8n.pt')
model.train(data='coco128.yaml', epochs=3, device=0, workers=1)
model.export(format='onnx', simplify=True, opset=12)
```
And the examplified inference part is as follows:
```cpp
// setup
cv::Mat frame; // net input image
cv::Size frameShape; // net input shape
std::string modelFile; // exported onnx model

cv::dnn::Net net = cv::dnn::readNet(modelFile);
net.setPreferableBackend(cv::dnn::DNN_BACKEND_CUDA);
net.setPreferableTarget(cv::dnn::DNN_TARGET_CUDA);
cv::dnn::Net* net_p = new cv::dnn::Net(net);

// inference
cv::Mat blob = cv::dnn::blobFromImage(frame, 1. / 255., frameShape, cv::Scalar(), true, false, CV_32F);
net_p->setInput(blob);
std::vector<cv::Mat> prob;
net_p->forward(prob, net_p->getUnconnectedOutLayersNames());

int prob_rows = prob[0].size[2];
int prob_cols = prob[0].size[1];
prob[0] = prob[0].reshape(1, prob_cols);
cv::transpose(prob[0], prob[0]);

float ratioh = (float)frame.rows / frameShape.height;
float ratiow = (float)frame.cols / frameShape.width;
float* data_p = (float*)prob[0].data;

std::vector<float> confidences;
std::vector<cv::Rect> boxes;
std::vector<int> classIds;
float confThreshold = 0.25f;
float nmsThreshold = 0.45f;

for (int n = 0; n < prob_rows; n++)
{
	cv::Mat scores = prob[0].row(n).colRange(4, prob_cols); // the scores are ok
	cv::Point classIdPoint;
	double maxClassScore;
	minMaxLoc(scores, 0, &maxClassScore, 0, &classIdPoint);
	if (maxClassScore >= confThreshold)
	{
		// the issue is observed here with a cuda build
		float cx = data_p[0] * ratiow; // data_p[0] = 0
		float cy = data_p[1] * ratioh; // data_p[1] = 0
		float w = data_p[2] * ratiow; // data_p[2] = 0
		float h = data_p[3] * ratioh; // data_p[3] = 0

		int left = int(cx - 0.5f * w);
		int top = int(cy - 0.5f * h);

		confidences.push_back((float)maxClassScore);
		boxes.push_back(cv::Rect(left, top, (int)(w), (int)(h)));
		classIds.push_back(classIdPoint.x);
	}
	data_p += prob_cols;
}

std::vector<int> indices;
cv::dnn::NMSBoxes(boxes, confidences, confThreshold, nmsThreshold, indices);
// visualize detection boxes here

// cleanup
delete net_p;
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by Krugpelke** [comment]
Also, I've opened a similar issue on the ultralytics YOLOv8 github as I wasn't sure which place was more appropriate to post to, let me know if this is okay.

**Comment by asmorkalov** [comment]
@Krugpelke Thanks for the issue report. Could you make small step back to cudnn 7.x and try again. There were some CuDNN 8.x integration issues in OpenCV before.

**Comment by Krugpelke** [comment]
Right, I've tested a build with cuda 10.2 as it seems to be the last release supported by the latest 7.6.5 cudnn version, but I'm getting the same results there.

In the meantime, I've checked layer outputs manually trying to find where I'm getting zeroes from, here's the end of yolov8's onnx graph, multiplication are bounding boxes and sigmoid are confidences.
![image](https://github.com/opencv/opencv/assets/60602483/5afcdc23-d066-4730-85ed-70a9a85ee9ff)
```
"431" -> "onnx_node!Mul_269"
"onnx_node!Mul_269" -> "onnx_node!Concat_271"
"onnx_node!Sigmoid_270" -> "onnx_node!Concat_271"
"onnx_node!Concat_271" -> "output0"
```
Now, I'm not sure what's going on here, running a forward pass on Mul_269 layer actually gives me non-zero values, however, my guess is that these are pre-multiplication, as they don't produce correct boxes
```c++
std::vector<cv::String> layerNames = { "onnx_node!Mul_269" };
std::vector<cv::Mat> prob;
net_p->forward(prob, layerNames);
// e.g. prob[0] here has non-zero values
// 3.412499, 20.992088, 28.306988, 38.623158, 47.723999, 51.463470, 55.425812, 57.841465, 58.026161...
```
Running a forward pass on both Mul_269 and Sigmoid_270 layers gives me zero values, the same as output0
```c++
std::vector<cv::String> layerNames = { "onnx_node!Mul_269", "onnx_node!Sigmoid_270" };
std::vector<cv::Mat> prob;
net_p->forward(prob, layerNames);
// e.g. prob[0] here has zero values
// 0.000000, 0.000000, 0.000000, 0.000000, 0.000000, 0.000000, 0.000000, 0.000000, 0.000000...
```
I've also noted that running a forward pass only on Sigmoid_270 layer produces an exception
```c++
std::vector<cv::String> layerNames = { "onnx_node!Sigmoid_270" };
std::vector<cv::Mat> prob;
net_p->forward(prob, layerNames);
// OpenCV(4.8.0) D:\opencv\opencv-4.8.0\modules\dnn\src\layers\../op_cuda.hpp:366: error: (-215:Assertion failed) offset == 0 in function 'cv::dnn::GenericCUDABackendWrapper<float,6>::copyToHost'
```
I wanted to check if the non-zero values could give me true bounding boxes by manually multiplying "430" and "431" layers, but I don't know how to do this multiplication correctly, as "430" is 1x4x8400 and "431" is 1x8400x1.

**Comment by Krugpelke** [comment]
Okay, I was able to get the correct bounding boxes from Mul_269, while testing I was reshaping the output blob differently, which I guess was messing with the element order.
```c++
// while testing I was doing this
prob[0] = prob[0].reshape(1, prob[0].size[2]);
// I should've been doing this
prob[0] = prob[0].reshape(1, prob[0].size[1]);
cv::transpose(prob[0], prob[0]);
```
So, the good news is that I've essentially managed to run yolov8 with cuda, however, this solution is a bit too cheeky: I'm running two separate forward passes on Mul_269 and the output layer and then I'm replacing the bounding boxes with the ones computed from Mul_269. If I could run a single forward pass on those layers here, I guess it actually wouldn't be that bad.
```c++
std::vector<cv::Mat> prob;
cv::Mat output432, output0;
net_p->forward(output432, "onnx_node!Mul_269");
net_p->forward(output0, "output0");
prob.push_back(output432);
prob.push_back(output0);
int prob_rows = prob[1].size[2];
int prob_cols = prob[1].size[1];
for (int i = 0; i < prob.size(); i++)
{
        prob[i] = prob[i].reshape(1, prob[i].size[1]);
        cv::transpose(prob[i], prob[i]);
}
float* bbox_data_p = (float*)prob[0].data;
for (int n = 0; n < prob_rows; n++)
{
	cv::Mat scores = prob[1].row(n).colRange(4, prob_cols); // the scores are ok
	cv::Point classIdPoint;
	double maxClassScore;
	minMaxLoc(scores, 0, &maxClassScore, 0, &classIdPoint);
	if (maxClassScore >= confThreshold)
	{
		float cx = bbox_data_p[0] * ratiow;
		float cy = bbox_data_p[1] * ratioh;
		float w = bbox_data_p[2] * ratiow;
		float h = bbox_data_p[3] * ratioh;
		// add confidences, boxes, ids here
	}
	bbox_data_p += 4;
}
```

**Comment by baudneo** [comment]
Do you know how the bbox data is getting zerod out? is this fix the only way to solve this issue ATM?

**Comment by baudneo** [comment]
I had to end up using onnxruntime for yolov8 and yolo-nas. The code here is passed the output of `net.forward()`

```python3
    def process_output(self, output):

        if output is not None:
            num_outputs = len(output)
            # YOLOv8
            if num_outputs == 1:
                predictions = np.squeeze(output[0]).T
                scores = np.max(predictions[:, 4:], axis=1)
                if len(scores) == 0:
                    # empty results
                    return np.empty((0, 4)), np.empty((0,)), np.empty((0,))

                # Get the class with the highest confidence
                # Get bounding boxes for each object
                boxes = self.extract_boxes(predictions)
                class_ids = np.argmax(predictions[:, 4:], axis=1)

            # YOLO-NAS
            elif num_outputs == 2:
                boxes: np.ndarray
                raw_scores: np.ndarray
                boxes, raw_scores = output
                scores = raw_scores.max(axis=2).flatten()
                if len(scores) == 0:
                    return np.empty((0, 4)), np.empty((0,)), np.empty((0,))

                boxes = np.squeeze(boxes, 0)
                boxes = self.rescale_boxes(boxes)
                class_ids = np.argmax(raw_scores, axis=2).flatten()
        else:
            return np.empty((0, 4)), np.empty((0,)), np.empty((0,))

        indices = cv2.dnn.NMSBoxes(
            boxes, scores, self.options.confidence, self.options.nms
        )
        return boxes[indices].astype(np.int32).tolist(), scores[indices].astype(np.float32).tolist(), class_ids[
            indices].astype(np.int32).tolist()
```

**Comment by mengqingmeng** [comment]
same issue, is there any way to slove?

**Comment by baudneo** [comment]
I switched to onnxruntime and it works perfectly. I recommend you to do the same.

**Comment by mengqingmeng** [comment]
> I switched to onnxruntime and it works perfectly. I recommend you to do the same.

thanks a lot, I will try later

**Comment by zhaoning999** [comment]
I test opencv 4.7.0, it's ok.
